core/store/migrations/migration1597695690/migrate.go

```diff
@@ -0,0 +1,194 @@
+            dot_dag_source TEXT NOT NULL,
+            created_at timestamptz NOT NULL
+        );
+
+        CREATE INDEX idx_pipeline_specs_created_at ON pipeline_specs USING BRIN (created_at);
+
+        CREATE TABLE pipeline_task_specs (
+            id SERIAL PRIMARY KEY,
+            dot_id TEXT,
+            pipeline_spec_id INT NOT NULL REFERENCES pipeline_specs (id) ON DELETE CASCADE,
+            type TEXT NOT NULL,
+            json jsonb NOT NULL,
+            index INT NOT NULL DEFAULT 0,
+            successor_id INT REFERENCES pipeline_task_specs (id),
+            created_at timestamptz NOT NULL
+        );
+
+        CREATE INDEX idx_pipeline_task_specs_pipeline_spec_id ON pipeline_task_specs (pipeline_spec_id);
+        CREATE INDEX idx_pipeline_task_specs_successor_id ON pipeline_task_specs (successor_id);
+        CREATE INDEX idx_pipeline_task_specs_created_at ON pipeline_task_specs USING BRIN (created_at);
+
+        CREATE TABLE pipeline_runs (
+            id BIGSERIAL PRIMARY KEY,
+            pipeline_spec_id BIGINT NOT NULL REFERENCES pipeline_specs (id) ON DELETE CASCADE,
+            meta jsonb,
+            created_at timestamptz NOT NULL
+            -- NOTE: Could denormalize here with finished_at/output/error of last task_run if that proves necessary for performance
+        );
+
+        CREATE INDEX idx_pipeline_runs_pipeline_spec_id ON pipeline_runs (pipeline_spec_id);
+        CREATE INDEX idx_pipeline_runs_created_at ON pipeline_runs USING BRIN (created_at);
+
+        CREATE TABLE pipeline_task_runs (
+            id BIGSERIAL PRIMARY KEY,
+            dot_id TEXT,
+            pipeline_run_id BIGINT NOT NULL REFERENCES pipeline_runs (id) ON DELETE CASCADE,
+            output JSONB,
+            error TEXT,
+            pipeline_task_spec_id BIGINT NOT NULL REFERENCES pipeline_task_specs (id) ON DELETE CASCADE,
+            index INT NOT NULL DEFAULT 0,
+            created_at timestamptz NOT NULL,
+            finished_at timestamptz,
+            CONSTRAINT chk_pipeline_task_run_fsm CHECK (
+                error IS NULL AND output IS NULL AND finished_at IS NULL
+                OR
+                error IS NULL AND output IS NOT NULL AND finished_at IS NOT NULL
+                OR
+                output IS NULL AND error IS NOT NULL AND finished_at IS NOT NULL
+            )
+        );
+
+
+        ---
+        --- Notify the Chainlink node when a new pipeline run has started
+        ---
+
+        CREATE OR REPLACE FUNCTION notifyPipelineRunStarted() RETURNS TRIGGER AS $_$
+        BEGIN
+            PERFORM pg_notify('pipeline_run_started', NEW.id::text);
+            RETURN NEW;
+        END
+        $_$ LANGUAGE 'plpgsql';
+
+        CREATE TRIGGER notify_pipeline_run_started
+        AFTER INSERT ON pipeline_runs
+        FOR EACH ROW EXECUTE PROCEDURE notifyPipelineRunStarted();
+
+
+        ---
+        --- Notify the Chainlink node when a pipeline run has completed
```

Comment:
Doing the trigger in SQL like this makes me nervous, it can be unexpectedly expensive and it's a completely hidden cost making it very difficult to debug performance problems down the line.

What about doing a manual NOTIFY in software instead? I mean, calling it directly in the go code in the pipeline runner since we already know there if a pipeline run finished or not.

The SQL is simply:

```go
tx.Exec(`NOTIFY pipeline_run_completed, ?::text`, pipelineRunID)
```

Reply:
Works for me, I had it like this before anyhow.  IIRC it requires `N + 1` extra queries (a SELECT after every PipelineTaskRun to see if the entire run is indeed finished + the NOTIFY)